### DISCUSSION TITLE
Internationalization for client side components

### DISCUSSION THREAD

**Discussion by llhooman** [vercel/next.js#57405] [App Router] [answered]
I am using nextjs version 13.5.6 and using app router Internationalization for translating my components,
but i get an error that its a server only feature and i can not translate my client side components , what am i doing wrong?

**Comment by landsman** [comment]
did you added "use client" in files where is client side code?

**Reply by llhooman** [reply]
i do, i get this error
```typescript
./lib\dictionary.ts
ReactServerComponentsError:

You're importing a component that needs server-only. That only works in a Server Component but one of its parents is marked with "use client", so it's a Client Component.
Learn more: https://nextjs.org/docs/getting-started/react-essentials

   ╭─[C:\Users\deghat\Desktop\Dana\dana\lib\dictionary.ts:1:1]
 1 │ import 'server-only'
   · ────────────────────
 2 │ import { Locale } from '../i18n.config'
 3 │  
 3 │ const dictionaries = {
   ╰────

One of these is marked as a client entry with "use client":
  ./lib\dictionary.ts
  ./src\app\[lang]\components\Typography\Button.tsx
```

**Comment by mass2527** [comment, marked as answer]
After adding the import statement import `'server-only'` in the `lib/dictionary.ts` file, all the code in that file is now designated as server-only code and cannot be imported from a client component.

If you need to access the dictionary from a client component, you should call the server-only function `getDictionary` in the `layout.tsx`(server component) file and then pass its result to a provider.

For example:


```tsx
// app/[lang]/layout.tsx

export default async function RootLayout({
  children,
  params,
}: {
  children: React.ReactNode
  params: { lang: Locale }
}) {
  const dictionary = await getDictionary(params.lang)

  return (
    <html lang={params.lang}>
      <body>
        <DictionaryProvider dictionary={dictionary}>
          {children}
        </DictionaryProvider>
      </body>
    </html>
  )
}
```

```tsx
// dictionary-provider.tsx
'use client'

type Dictionary = Awaited<ReturnType<typeof getDictionary>>

const DictionaryContext = React.createContext<Dictionary | null>(null)

export default function DictionaryProvider({
  dictionary,
  children,
}: {
  dictionary: Dictionary
  children: React.ReactNode
}) {
  return (
    <DictionaryContext.Provider value={dictionary}>
      {children}
    </DictionaryContext.Provider>
  )
}

export function useDictionary() {
  const dictionary = React.useContext(DictionaryContext)
  if (dictionary === null) {
    throw new Error('useDictionary hook must be used within DictionaryProvider')
  }

  return dictionary
}
```

```tsx
'use client'

export default function YourClientComponent(){
   const dictionary = useDictionary()
   return <div>{dictionary.something}</div>
}
```

**Reply by llhooman** [reply]
Thank you this fixed my issue!

**Reply by hichui** [reply]
Hi, i'm doing  exactly the same but i get this error : _"Functions cannot be passed directly to Client Components unless you explicitly expose it by marking it with "use server"."_ It's referring to the dictionary var that we pass to the provider. Can you help me please? thank you

**Reply by mass2527** [reply]
@hichui 
With the information provided, I'm unable to assist in resolving the issue. Could I please review your code?

**Reply by beforedisappear** [reply]
Some kind of meaningless implementation. Then why "server only"? And if the dictionary is huge, put it all in context?

**Reply by Haeyzed** [reply]
it worked for me.

**Reply by krckyboy** [reply]
As @beforedisappear said, it doesn't make sense to put the entire tree in context for reasons below.

I literally created something very similar to this just today, where I fetch translations in `layout.tsx` via server action, and then pass it down into translation provider so I can easily get access to it in client components. 

This would work perfectly, if Next.js wasn't overpopulating `<head>` with `<scripts>`, and this is a major concern for production. 

This way we expose our entire translation tree in head, and we are left off with a big hit on performance due to enormous size of HTML.

It's just a tiny part:

![image](https://github.com/vercel/next.js/assets/11292329/5e3b97a1-43d2-4436-b04b-b3cc03e3a2fd)

**Reply by ninaismail** [reply]
Bear with me please I have 2 questions: 1) why can't we make getDictionary() not a server only function? And 2) should I use the data from the DictionaryProvider in my client components, and the one from server-only in my server components?
